File: cdap-api/src/main/java/co/cask/cdap/api/dataset/lib/File.java
/*
 * Copyright © 2014 Cask Data, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

package co.cask.cdap.api.dataset.lib;

import co.cask.cdap.api.annotation.Beta;
import co.cask.cdap.api.data.batch.InputFormatProvider;
import co.cask.cdap.api.data.batch.OutputFormatProvider;
import co.cask.cdap.api.dataset.Dataset;
import org.apache.twill.filesystem.Location;

import java.util.List;

/**
 * This dataset represents a collection of files on the file system. The dataset has a base location, under which
 * all of its files are located. When instantiated, runtime arguments are required to specify the individual file
 * or files being used.
 */
@Beta
public interface File extends Dataset, InputFormatProvider, OutputFormatProvider {

  /**
   * The base path of the dataset.
   */
  String PROPERTY_BASE_PATH = "base.path";

  /**
   * The name of the input format class.
   */
  String PROPERTY_INPUT_FORMAT = "input.format";

  /**
   * The name of the output format class.
   */
  String PROPERTY_OUTPUT_FORMAT = "output.format";

  /**
   * Prefix for additional properties for the input format. They are added to the
   * Hadoop configuration, with the prefix stripped from the name.
   */
  String PROPERTY_INPUT_PROPERTIES_PREFIX = "input.properties.";

  /**
   * Prefix for additional properties for the output format. They are added to the
   * Hadoop configuration, with the prefix stripped from the name.
   */
  String PROPERTY_OUTPUT_PROPERTIES_PREFIX = "output.properties.";

  /**
   * The paths of the files to be read. Specified as a runtime argument for the dataset.
   * Each path is relative to the dataset's base path, and multiple paths can be given,
   * separated by commas.
   */
  String ARGUMENT_INPUT_PATHS = "input.paths";

  /**
   * The path of the file to write. Specified as a runtime argument for the dataset.
   * The path is relative to the dataset's base path.
   */
  String ARGUMENT_OUTPUT_PATH = "output.path";

  /**
   * Allows to interact directly with the location of this dataset in the underlying file system.
   *
   * @return the location of the base directory
   */
  Location getBaseLocation();

  /**
   * Allows direct access to files of this dataset in the underlying file system.
   *
   * @return the list of input locations
   */
  List<Location> getInputLocations();

  /**
   * Allows direct access to files in the output location, in the underlying file system.
   *
   * @return the output location
   */
  Location getOutputLocation();

  /**
   * Allows direct access to files in the input locations, in the underlying file system.
   *
   * @return the location of the given relative path within this dataset
   */
  Location getLocation(String relativePath);
}

File: cdap-ui/app/cdap/components/DataPrep/DataPrepBrowser/DataPrepBrowserStore/Actions/file.js
/*
 * Copyright © 2018 Cask Data, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */

import { setActiveBrowser, setError } from './commons';
import DataPrepBrowserStore, {
  Actions as BrowserStoreActions,
} from 'components/DataPrep/DataPrepBrowser/DataPrepBrowserStore';
import NamespaceStore from 'services/NamespaceStore';
import MyDataPrepApi from 'api/dataprep';
import { convertBytesToHumanReadable, HUMANREADABLESTORAGE_NODECIMAL } from 'services/helpers';
import uuidV4 from 'uuid/v4';
import moment from 'moment';
import T from 'i18n-react';
const PREFIX = 'features.FileBrowser';

const trimSuffixSlash = (path) => path.replace(/\/\//, '/');

const formatResponse = (contents) => {
  return contents.map((content) => {
    content.uniqueId = uuidV4();
    content['last-modified'] = moment(content['last-modified']).format('MM/DD/YY HH:mm');
    content.displaySize = convertBytesToHumanReadable(
      content.size,
      HUMANREADABLESTORAGE_NODECIMAL,
      true
    );

    if (content.directory) {
      content.type = T.translate(`${PREFIX}.directory`);
    }
    content.type = content.type === 'UNKNOWN' ? '--' : content.type;

    return content;
  });
};

const goToPath = (path) => {
  path = trimSuffixSlash(path);
  setFileSystemLoading();
  setFileSystemPath(path);

  const namespace = NamespaceStore.getState().selectedNamespace;

  MyDataPrepApi.explorer({
    context: namespace,
    path,
    hidden: true,
  }).subscribe(
    (res) => {
      DataPrepBrowserStore.dispatch({
        type: BrowserStoreActions.SET_FILE_SYSTEM_CONTENTS,
        payload: {
          contents: formatResponse(res.values),
        },
      });
    },
    (err) => {
      setError(err);
    }
  );
};

const setFileSystemLoading = () => {
  DataPrepBrowserStore.dispatch({
    type: BrowserStoreActions.SET_FILE_SYSTEM_LOADING,
    payload: {
      loading: true,
    },
  });
};

const setFileSystemAsActiveBrowser = (payload) => {
  setActiveBrowser(payload);
  goToPath(payload.path);
};

const setFileSystemPath = (path) => {
  DataPrepBrowserStore.dispatch({
    type: BrowserStoreActions.SET_FILE_SYSTEM_PATH,
    payload: {
      path,
    },
  });
};

const setFileSystemSearch = (search) => {
  DataPrepBrowserStore.dispatch({
    type: BrowserStoreActions.SET_FILE_SYSTEM_SEARCH,
    payload: {
      search,
    },
  });
};

export {
  goToPath,
  trimSuffixSlash,
  setFileSystemLoading,
  setFileSystemAsActiveBrowser,
  setFileSystemSearch,
};
